Support draft assemblies

muffato opened the pull request:
On this branch, there is no input Yaml file. The only mandatory parameters are:

- Species name / taxon_id (`--taxon`)
- Assembly (`--fasta`)
- Sample sheet (`--input`) to list the read files

`--accession` is optional and is used to pull assembly information from ENA into the blobDir's meta.json.

I haven't restructured the pipeline much. All the blobtools command at the end still require a yaml file. My solution is to add a script at the beginning of the pipeline that generates the minimal yaml file required (as per https://github.com/sanger-tol/blobtoolkit/issues/77#issuecomment-1936286274). It still allows clearly getting some parameters in the input-check sub-workflow and making the busco sub-workflow more focused on running buco + blastp.

Busco lineages are inferred from the taxonomy directly here. Like in the genome-note pipeline, I've moved away from using GoaT as GoaT is just a proxy to the NCBI taxonomy. This way, I can keep control of both the version of Busco and the list of lineages in the same place.
I've also introduced the `--busco_lineages` parameter to allow precisely selecting the lineages that are used, rather than the taxonomy-based defaults.

Still a draft for now as I want to review `/nfs/team135/yy5/btk_config/taxonomiser_v2.py` and maybe incorporate some elements of it.

<!--
# sanger-tol/blobtoolkit pull request

Many thanks for contributing to sanger-tol/blobtoolkit!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](.github/CONTRIBUTING.md)
-->

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

muffato commented:
I've added some code to achieve the goal of `taxonomiser_v2.py`, which is: find a taxon_id that is recognised by the NT database and the closest to the species of interest.
It's implemented very differently from the script. I leverage the `taxonomy4blast.sqlite3` database that is shipped with NT and essentially lists the taxon_ids it knows about. If the species' taxon_id is not recognised, then it looks for the parent, etc.

As far as I understand the requirements, this is the last bit that was missing to complete support for draft assemblies. I'll mark this pull-request as ready.

muffato commented:
@eeaunin . I've rebased this branch. It now includes the fixes I've made for blast

eeaunin commented on docs/usage.md:
```diff
@@ -189,9 +183,13 @@ Nextflow
 
 # Draft Assemblies
-nextflow run sanger-tol/blobtoolkit --input SAMPLESHEET --fasta GENOME –-accession TAG --taxon TAXON_ID --yaml CONFIG --taxdump TAXDUMP_DB --blastp DMND_db --blastn BLASTN_DB --blastx BLASTX_DB
+nextflow run sanger-tol/blobtoolkit --input SAMPLESHEET --fasta GENOME --taxon TAXON_ID --taxdump TAXDUMP_DB --blastp DMND_db --blastn BLASTN_DB --blastx BLASTX_DB
```
It looks like the taxID given by the user is used as the argument for `-negative_taxids` in the downstream `blastn` command. I assume the purpose of `-negative_taxids` is to ignore hits to the same assembly itself if the assembly is already in the `nt` database. I think this makes sense to do when running BlobToolKit with already published assemblies but not when running it with draft assemblies.

When testing `sanger-tol/blobtoolkit` I ran into a `blastn` crash that seems to originate from the `-negative_taxids` part in the BLAST command: 
```
BLAST Database error: Taxonomy ID(s) not found.Taxonomy ID(s) not found. This could be because the ID(s) provided are not at or below the species 
level. Please use get_species_taxids.sh to get taxids for nodes higher than species (see https://www.ncbi.nlm.nih.gov/books/NBK546209/).
```
The taxID that I used was 73239. This actually already is at subspecies level but it somehow still triggered this error. I think this crash wouldn't have happened if  the run with the draft assembly hadn't required the taxID for the `-negative_taxids` bit. So I think it would be better to remove the `-negative_taxids` part when the run is done with a draft assembly

muffato replied:
For comparison, can you share an example run of the Snakemake pipeline ? I was under the impression that `-negative_taxids` was _always_ used.

This "BLAST Database error: Taxonomy ID(s) not found." error has been fixed in #103, which is included in this branch since my last rebase 2 weeks ago. Are you still seeing the error on the latest version of the branch ?

eeaunin replied:
About the `blastn` crash: yes, sorry, turns out I was accidentally running out of date code when I had this crash yesterday. The crash didn't happen when I retried the run with updated repository copy.

eeaunin commented:
I had a closer look at how `-negative_taxids` has been implemented in the Snakemake pipeline and it appears quite confusing. The BlobToolKit paper ([https://academic.oup.com/g3journal/article/10/4/1361/6026202](https://academic.oup.com/g3journal/article/10/4/1361/6026202)) says:

> An optional filter excludes a configurable list of NCBI taxIDs (default: excludes query genus).

So the exclusion of taxids is supposed to be optional and configurable by the user.
BlobToolKit pipeline v1 has the `mask_ids` setting for excluding taxids:

[https://github.com/blobtoolkit/pipeline/blob/master/v1/example.yaml](https://github.com/blobtoolkit/pipeline/blob/master/v1/example.yaml)

However, I couldn't find a setting for the same thing in the Snakemake pipeline v2 code. Maybe the authors just forgot to include it?

In my runs with the Snakemake pipeline negative taxids were not used but there are suppressed error messages buried in the run logs relating to that. In a run with a Plasmodium yoelii yoelii assembly there is this error in the logs (`/lustre/scratch123/tol/teams/tola/users/ea10/pipeline_testing/20230215_pyoelii_asg_cobiont_check_run/btk_busco/blastn/logs/pyoelii/run_blastn.log`):

```
BLAST Database error: Taxonomy ID(s) not found.Taxonomy ID(s) not found. This could be because the ID(s) provided are not at or below the species level. Please use get_species_taxids.sh to get taxids for nodes higher than species (see https://www.ncbi.nlm.nih.gov/books/NBK546209/).
Restarting blastn without taxid filter
```

So it ran into the error but then just quietly continued running. It is unclear to me what caused this error, as the taxid used there (352914) is at strain level.

In another run it has skipped using the taxid filter due to another error: `/lustre/scratch123/tol/teams/grit/contamination_screen/icMagCera1/20240712_icMagCera1.20240711.hap1.fa_asg_cobiont_check_run/btk_busco/blastn/logs/icMagCera1.20240711.hap1.fa/run_blastn.log`

```
BLAST Database error: Taxonomy filtering is not supported in v4 BLAST dbs
Restarting blastn without taxid filter
```

So the filtering doesn't work if the supplied database is V4 instead of V5 but this also doesn't crash the Snakemake pipeline and just produces an error message in the logs.

I guess it would be okay if the `sanger-tol/blobtoolkit` pipeline used `-negative_taxids` in all runs with draft assemblies as long as this doesn't produce frequent crashes. But I think it would be better if the use of `-negative_taxids` was optional for draft assemblies.

muffato commented:
@eeaunin . I've added a `--skip_taxon_filtering` for you. It removes the taxon filtering from all Blast searches

I've rebased the branch onto the latest stable release 0.5.1

eeaunin commented:
That's good then! I think it's fine to merge the `draft_assemblies` branch to `dev` now
